tiledb/sm/query_plan/query_plan.cc

```diff
@@ -50,51 +50,50 @@ namespace sm {
 /* ********************************* */
 QueryPlan::QueryPlan(Query& query) {
   if (query.array()->is_remote()) {
-    throw std::logic_error(
-        "Failed to create a query plan; Remote arrays"
-        "are not currently supported.");
-  }
+    auto rest_client = query.rest_client();
+    if (!rest_client) {
+      throw std::runtime_error(
+          "Failed to create a query plan; Remote query"
+          "with no REST client.");
+    }
 
-  array_uri_ = query.array()->array_uri().to_string();
-  vfs_backend_ = URI(array_uri_).backend_name();
-  query_layout_ = query.layout();
+    query_plan_ = rest_client->post_query_plan_from_rest(
+        query.array()->array_uri(), query);
+    // We need to transition the query status to INITIALIZED to mimic the
+    // behavior of getting a query plan locally
+    query.set_status(QueryStatus::INITIALIZED);
```

Comment:
What was the logic behind this? It seems very much like we're covering either a bug or a weird test. But just setting the state to initialized seems not correct on the face of it.

If we actually do need this behavior to match state, I'd call `query.strategy()` to mimic the local side. But I also lean towards not doing either because it seems really weird that getting a QueryPlan somehow modifies the query in such a way that something else breaks by relying on the QueryPlan's constructor behavior modifying the query state.

Reply:
I 100% agree with you on the weirdness of this. I thought the same when I saw QueryPlan having side effects on Query, but that's how QueryPlan feature was designed. If you prefer that I call `query.strategy()`, I can do that easily I guess, but if we talk about removing this completely then we should rethink and change the `QueryPlan` design. 
I guess @robertbindar can also chime in here.

Reply:
@ypatia thanks for tagging. Transitioning the query to a new state was a requirement of the query plan design. A query plan doesn't make much sense without a fully formed query, but we don't yet have a C41 query which would generate its plan during construction. The `INITIALIZED` intermediate state was **a** solution to simulate a fully formed query (`INITIALIZED` makes the query immutable in a way) without attempting to make the `Query` class C41 and at the same time hope it'll reduce debt buildup until we make the move on the `Query` class.

`query.strategy()` doesn't make sense on the client, calling it might break things or simply it could just create the incorrect strategy (not necessarily now, but probably in the future). Setting the query state to `INITIALIZED` here on the client is just a way to make the query immutable as it would be if the array would have been local.